feat: Terrain variation

davo0411 opened the pull request:
Final Version.
Distance based effect, only applies at a set range controlled by sliders in settings.
Lower perf. cost than before because less effect.
Works without plugin installed. Works with latest dev. Works with terrain Helper & VR.

![AfterTundra](https://github.com/user-attachments/assets/9b8250a8-c879-47b4-afd6-a2cf2d63d7e6)
![BeforeTundra](https://github.com/user-attachments/assets/c31ba113-0f79-4999-a8a6-edda09686dc0)

davo0411 committed:
Update Lighting.hlsl

davo0411 committed:
Added a Terrain Tiling Fix to Lighting.hlsl

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Update lighting.hlsl

Fixed small compile issue with PBR. No difference to visual effect.

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
First of several fixes

davo0411 committed:
Merge branch 'terrain-variation' of https://github.com/davo0411/skyrim-community-shaders into terrain-variation

davo0411 committed:
Update Lighting.hlsl

davo0411 committed:
Separate Feature Beta 1

Working Terrain Tiling Fix as a separate feature within Community Shaders. Has a on/off switch in the settings.

davo0411 committed:
Updated Split plugin

Works with or without terrainvariation plugin enabled now

davo0411 committed:
Update

davo0411 committed:
Fix for Parallax Shadows on Whole Terrain

Fix for weird issue with parallax terrain shadows causing the whole terrain to go black at certain distances.

davo0411 committed:
FINAL

Plugin now works fully when installed, and does not throw any errors when TerrainVariation.hlsli is not present. This allows the plugin to be a separate nexus release.

davo0411 committed:
Comment Cleanup 1

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Delete .gitignore

davo0411 committed:
Update ExtendedMaterials.hlsli

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
fixing gitignore

davo0411 committed:
Updating References

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Name Fix, General Tidy-up

davo0411 committed:
Optimisations 1

Added checks for land weights, ensuring that sampling doesn't occur if it isn't needed. Should reduce overhead. Also tidied up the code, incorporating the blending of diffuse/rmaos/alpha into the layer code.

davo0411 committed:
Optimisations 1.1

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Attempt 1 to support terrain helper

davo0411 committed:
Merge branch 'dev' into Terrain-Variation-ASFEATURE

davo0411 committed:
Merge branch 'dev' into Terrain-Variation-ASFEATURE

davo0411 committed:
Parallax de-borking

Basically reverting a lot of stuff i broke while testing back to original code

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Tidy up

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Setting up settings

davo0411 committed:
Update TerrainVariation.cpp

davo0411 committed:
fix

davo0411 committed:
Compile Fix

NOW WORKS WITH PARALLAX & PARALLAX SHADOWS ONLY. NOT SAMPLING NORMAL YET.

davo0411 committed:
Update, working normals

davo0411 committed:
Get parallax and shadows working

davo0411 committed:
DARKER SHADOWS BIGGER PARALLAX

davo0411 committed:
User-Definable Parallax and P Shadow Height Changes

davo0411 committed:
Fix for the parallax controls

davo0411 committed:
Merge branch 'doodlum:dev' into Terrain-Variation-Distancebased

davo0411 committed:
Capitalisation fixes.

davo0411 committed:
Dynamic Contrast fix tied to distance

davo0411 committed:
Clean up

davo0411 committed:
no more swimmy textures

davo0411 committed:
comment and defaults tidy

davo0411 committed:
Update StochasticSampling.hlsli

davo0411 committed:
MAJOR REWRITE PART 1

davo0411 committed:
benableimprovedsnow fixes

davo0411 committed:
fix

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Comment & Code cleanup

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
adding haka's nexus download stuff

davo0411 committed:
Merge branch 'Terrain-Variation-Distancebased' of https://github.com/davo0411/skyrim-community-shaders into Terrain-Variation-Distancebased

davo0411 committed:
style: 🎨 apply clang-format changes

doodlum reviewed (changes requested):
Getting a massive performance hit from this. 15% hit to my GPU (RTX 4080)

davo0411 committed:
doodlez fixes 1

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Restored distance-based effect

davo0411 committed:
Faster HASH2D2D Function.

Removed fmod, faster function now. Looks indistinguishable

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
undo

davo0411 committed:
SharedOffset.

Majorly improves performance as there is now one shared offset across all LandBlendWeights.

davo0411 committed:
Merge branch 'Terrain-Variation-Distancebased' of https://github.com/davo0411/skyrim-community-shaders into Terrain-Variation-Distancebased

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Landblendweights > 0.01

davo0411 committed:
Merge branch 'Terrain-Variation-Distancebased' of https://github.com/davo0411/skyrim-community-shaders into Terrain-Variation-Distancebased

davo0411 committed:
Added Quality Settings, High/low

davo0411 committed:
Optimisations

davo0411 committed:
Better low quality mode

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Misc fixes

davo0411 committed:
Merge branch 'Terrain-Variation-Distancebased' of https://github.com/davo0411/skyrim-community-shaders into Terrain-Variation-Distancebased

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
minor optimise

davo0411 committed:
Merge branch 'Terrain-Variation-Distancebased' of https://github.com/davo0411/skyrim-community-shaders into Terrain-Variation-Distancebased

davo0411 committed:
fix

davo0411 committed:
no more fail message

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
fix for drawfailloadmessage

davo0411 commented:
adding lod support & fixing bug with low mipmap terrain using wrong weights, made draft for now

doodlum reviewed (changes requested):
Extended Materials will not compile unless Terrain Variation is enabled. This needs to be fixed. As well as branch conflicts.

davo0411 committed:
Merge branch 'dev' into Terrain-Variation-Distancebased

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
fix

davo0411 committed:
Merge branch 'Terrain-Variation-Distancebased' of https://github.com/davo0411/skyrim-community-shaders into Terrain-Variation-Distancebased

davo0411 committed:
fixed merge for CS Hair

davo0411 committed:
Update Lighting.hlsl

davo0411 committed:
float viewdistance fix

davo0411 committed:
update

davo0411 committed:
holy shit finally fixed this sucked

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 commented:
Branch conflicts fixed, could not replicate extendedmaterials issue. Nothing in there has been touched for a while, tested with tv and without and worked fine.

alandtse commented on features/Extended Materials/Shaders/ExtendedMaterials/ExtendedMaterials.hlsli:
```diff
@@ -448,39 +600,80 @@ namespace ExtendedMaterials
 																																			max(params[3].HeightScale * input.LandBlendWeights1.w, max(params[4].HeightScale * input.LandBlendWeights2.x, params[5].HeightScale * input.LandBlendWeights2.y)))));
 			if (scale < 0.01)
-				return 1.0;
+				if (scale < 0.01)
```
Is this redundant of the line above?

davo0411 replied:
Typo, thank you! Fixed.

alandtse replied:
I don't want to be harsh but I'm pretty sure this would've broken in a compiler. It doesn't promote confidence about your testing if this made it through and you said it was compiling without issues.

The reality is this will delay a PRs merge when we're worried about things like this.

davo0411 replied:
It's alright, but yeah it wasn't having any impact I could see. Potentially taking 0.01% of the 0.01, but beyond that it didn't break anything and went through fine. Fixed now anyhow

alandtse commented on features/Terrain Variation/Shaders/Features/TerrainVariation.ini:
```diff
@@ -0,0 +1,2 @@
+[Info]
+Version = 1-0-1
```
New launches should be 1-0-0.

davo0411 replied:
Fixed

alandtse reviewed (changes requested):
Given how massive this is, the diff showing weird code, and the conflict issues you had, you'll need to confirm no hlsl warnings and errors for enabled, disabled, and disabled at boot. Also, you need a test build out too for testing. 

Hopefully this gets merged in early before other code gets added causing conflicts.

davo0411 committed:
Version fix

davo0411 commented:
> Given how massive this is, the diff showing weird code, and the conflict issues you had, you'll need to confirm no hlsl warnings and errors for enabled, disabled, and disabled at boot. Also, you need a test build out too for testing.
> 
> Hopefully this gets merged in early before other code gets added causing conflicts.

I have tested myself with enabled and disabled and saw 0 shader compile issues. Everything is contained within the applicable if def. I can test with it disabled at boot tonight. I will update this when i do.

I have a test build, Terrain Variation RC4. This is functionally identical to that but that is prior to yesterday's merges. I can create a new one but nothing in my code has changed, only the dev it is built upon. I verified everything worked as intended.

Pentalimbed commented on features/Terrain Variation/Shaders/TerrainVariation/TerrainVariation.hlsli:
```diff
@@ -0,0 +1,102 @@
+}
+
+// Hash function for stochastic sampling
+inline float2 hash2D2D(float2 s)
```
Put these in `Random.hlsli`

davo0411 replied:
These are only used in this code, its a relatively specific hash function. I can move if it i have to but i'd prefer to keep it consolidated here

Pentalimbed replied:
I mean it's a hash, how specific could it be? You can make two hash functions for each quality version and write some comments about quality/perf. Maybe someone else will try.

davo0411 replied:
Maybe specific is the wrong word, more pedantic ig. Feature fully breaks if these magic numbers are wrong in any way. Doesn't even just become worse, just fully stops working if its not multiplied by the right values/functions I've found. Perf version was as cheap as i could go without bork

Pentalimbed replied:
> Feature fully breaks if these magic numbers are wrong in any way.

That's basically every hash function tho.

alandtse replied:
I'm ok with this being a refactor to move to random after merge. 

But, I'm a bit freaked out that magic numbers breaks the feature if even changed a little. That doesn't sound right but I defer to others with better understanding of hashing functions.

davo0411 committed:
Typo Fix in ExtMat.hlsli

davo0411 committed:
style: 🎨 apply clang-format changes

Pentalimbed commented on package/Shaders/Common/SharedData.hlsli:
```diff
@@ -161,6 +161,17 @@ namespace SharedData
 		float pad;
 	};
 
+	struct TerrainVariationSettings
```
Pad the thing if it's going into a constant buffer

davo0411 replied:
Fixed, silly oversight

davo0411 committed:
padding

davo0411 committed:
Merge branch 'Terrain-Variation-Distancebased' of https://github.com/davo0411/skyrim-community-shaders into Terrain-Variation-Distancebased

alandtse commented:
> I have a test build, Terrain Variation RC4. This is functionally identical to that but that is prior to yesterday's merges. I can create a new one but nothing in my code has changed, only the dev it is built upon. I verified everything worked as intended.

Unfortunately, that RC4 is invalid now that you had conflicts to address. It must match the code in the PR.

davo0411 commented:
> > I have a test build, Terrain Variation RC4. This is functionally identical to that but that is prior to yesterday's merges. I can create a new one but nothing in my code has changed, only the dev it is built upon. I verified everything worked as intended.
> 
> Unfortunately, that RC4 is invalid now that you had conflicts to address. It must match the code in the PR.

No sweat, will post tonight (5-6 hours ish from now). Should be good timing for dood

davo0411 committed:
comment update

davo0411 commented:
Final Test build is up on discord. Feature is complete, no failed shaders (installed, uninstalled & disabled at startup), merged correctly with Jiyae's content.

doodlum reviewed (approved):
accepting this PR although I'm still not happy with the performance
